text/0000-interior.md

```diff
@@ -0,0 +1,525 @@
+trait, and which takes an instance of `Interior<Self>` by value:
+
+```rust
+#[lang="drop_interior"]
+trait DropInterior {
+    fn drop_interior(Interior<Self>);
+}
+```
+
+Define the following functions to convert between `T` and
+`Interior<T>`:
+
+```rust
+impl<T> Interior<T> {
+    unsafe fn from_outer_unsafe(outer: T) -> Self;
+    fn into_outer(this: Self) -> T;
+    fn as_outer(this: &Self) -> &T;
+    fn as_outer_mut(this: &mut Self) -> &mut T;
+    fn of_outer(outer: &T) -> &Self;
+    fn of_outer_mut(outer: &mut T) -> &mut Self;
+}
+```
+
+(Note that `from_outer_unsafe` is made unsafe to avoid a
+backwards-compatibility risk with current `Drop`. This design point
+will also be discussed under **Alternatives**.)
+
+These functions will be used to define conversion traits against
+`Interior<T>`:
+
+```rust
+impl<T> AsRef<T> for Interior<T> { ... }
+impl<T> AsMut<T> for Interior<T> { ... }
+```
+
+(With negative bounds, a safe variant of `from_outer` could be defined
+for !Drop types, and `From<T>` could be implemented for `Interior<T:
+!Drop>`. This is discussed under **Alternatives**.)
+
+With these definitions in place, it is possible to define the legacy
+`Drop` trait in terms of `DropInterior`:
+
+```rust
+impl<T: Drop> DropInterior for T {
+    fn drop_interior(mut this: Interior<T>) {
+        Interior::as_outer_mut(&mut this).drop();
+    }
+}
+```
+
+## Structurally identical
+
+`Interior<T>` is defined to be structurally identical to `T`. There
+are several things we mean by that:
+
+* `Interior<T>` will have the same size and alignment as `T`.
+* The fields of `Interior<T>` will have the same names as the fields of
+`T` and will live at the same memory offsets within the structure.
+* If `T` is an enum, then `Interior<T>` will have the same enum
+discriminant and variants.
+* The accessability of the fields of `Interior<T>` will be identical
+to the accessability of the equivalent fields of `T`.
+
+This definition is intended to support having the conversion from `T`
+to `Interior<T>` (and vice-versa) be a no-op at run-time: the
+introduction of the `Interior<T>` type should introduce *no* run-time
+overhead.
+
+## Drop-glue functions
+
+In this design, we have eliminated the need for the drop-glue function
+for a type to be compiler-generated. Rather, we make the drop-glue
+function (called on scope-based clean-up for a `DropGlue`-value)
+explicit, with an associated lang-item:
+
+```rust
+#[lang="drop_glue_fn"]
+fn drop_glue<T: DropInterior>(arg: T) {
+    // using "unsafe" variant to allow working with `Drop` types (as
+    // opposed to `DropInterior` types that do NOT implement `Drop`).
+    DropInterior::drop_interior(unsafe { Interior::from_outer_unsafe(arg) });
+}
+```
+
+(Note that, under this proposal, all drop-hooks are invoked via
+`DropInterior`, so restricting the `T` argument to implementors of
+`DropInterior` will cover all types that require `DropGlue`.) This
```

Comment:
@rkjnsn That is good and all, but it's incomplete.
Look at the definition of `DerefOwn` and the implementation of `DerefOwn` for `Box` and try to write it without `OwnIn` being generic over the "place" that manages the allocation of the value.
Remember, such a pointer always owns the *value*, but ownership and cleanup of the underlying memory may vary from case to case.

Your use of `in` is also ambiguous in parsing with the planned in-place construction syntax whether it's `in(place) expr` or `in place {expr}`.

At least with `in` you could try to accommodate the two together, but `out` cannot be used like that because it's not already a keyword, not even reserved.
We *do not* have a magical way to deal with unreserved keywords, except in contextual-keyword-friendly contexts (such as those where an item always starts with a keyword).
Expressions *are not* friendly to anything, really, they're quite nasty.
Simple counter-example: `let out = x; if out == &out {y} else {z}`.